app/models/smart_proxy_feature.rb

```diff
@@ -0,0 +1,30 @@
+class SmartProxyFeature < ApplicationRecord
+  belongs_to :smart_proxy
+  belongs_to :feature
+  validates :feature, :uniqueness => {:scope => :smart_proxy_id}
+
+  serialize :capabilities, Array
+  store :settings, coder: JSON
+
+  def details
+    {settings: settings, capabilities: capabilities}
+  end
+
+  def self.import_features(smart_proxy, features_json)
+    name_map = Feature.name_map
+    new_feature_classes = features_json.keys.map {|feature| name_map[feature]}
+    smart_proxy.smart_proxy_features.where.not(feature_id: new_feature_classes).destroy_all
+    features_json.each do |name, feature_json|
+      feature_class = name_map[name]
+      smart_proxy_feature = smart_proxy.smart_proxy_features.to_a.find {|spf| spf.feature_id == feature_class.id}
+      smart_proxy_feature ||= SmartProxyFeature.new(:feature_id => feature_class.id)
+      smart_proxy_feature.import_json(feature_json)
+      smart_proxy.smart_proxy_features << smart_proxy_feature
+    end
```

Comment:
how about:
```ruby
features_json.each do |name, feature_json|
  feature_class = name_map[name]
  smart_proxy_feature =  SmartProxyFeature.find_or_create_by(smart_proxy: smart_proxy, feature: feature_class)
  smart_proxy_feature.import_json(feature_json)
end

Reply:
That is clearer, but depends on all the smart_proxy_features having been saved to the db, #associate_features is called as a smart_proxy before_save so these aren't persisted yet to the database.  (Which means the find_or_create_by will not locate the existing feature, causing a duplicate to be associated)

I'd much prefer associate_features NOT be called as part of the active record creation, as that makes all sorts of things more tricky, but didn't want to tackle that as part of this PR.

Reply:
ok, that's reasonable, can you just add a comment there so the future me won't have an urge to refactor it :slightly_smiling_face: ?